Utils/CSV.cs

```diff
@@ -0,0 +1,409 @@
+				}
+				catch (FormatException)
+				{
+					throw new ArgumentException(string.Format("Input '{0}' cannot be parsed to type {1}", field, typeof(T)));
+				}
+				catch (ArgumentException)
+				{
+					throw new ArgumentException(string.Format("Input '{0}' is invalid for the constraints on this {1}", field, this.GetType()));
+				}
+			}
+		}
+
+		public static class Types
+		{
+			public class String : Type<string>
+			{
+				public String(bool allowEmptyValues = false) : base(s => s, allowEmptyValues) { }
+			}
+
+			public class Int : Type<int>
+			{
+				public Int() : base(int.Parse) { }
+
+				public Int(bool allowEmptyValues = false, int minValue = int.MinValue, int maxValue = int.MaxValue) : base(s => 
+				{
+					var value = int.Parse(s);
+					if (value < minValue) throw new ArgumentException("Min value not reached: " + maxValue);
+					if (value > maxValue) throw new ArgumentException("Max value exceeded: " + maxValue);
+					return value;
+				}, allowEmptyValues){}
+			}
+
+			public class Bool : Type<bool>
+			{
+				public Bool(bool allowEmptyValues = false) : base(bool.Parse, allowEmptyValues) { }
+			}
+
+			public class Float : Type<float>
+			{
+				public Float() : base(float.Parse) { }
+
+				public Float(bool allowEmptyValues = false, float minValue = float.MinValue, float maxValue = float.MaxValue) : base(s =>
+				{
+					var value = float.Parse(s);
+					if (value < minValue) throw new ArgumentException("Min value not reached: " + maxValue);
+					if (value > maxValue) throw new ArgumentException("Max value exceeded: " + maxValue);
+					return value;
+				}, allowEmptyValues){ }
+			}
+
+			public class Enum<T> : Type<T> where T : struct, IConvertible
+			{
+				public Enum(bool ignoreCase = false, bool allowEmptyValues = false, T fallbackValue = default(T)) : base(s =>
+				{
+					if (!typeof(T).IsEnum) throw new ArgumentException("Type must be an Enum type.");
+
+					if (string.IsNullOrEmpty(s))
+					{
+						if (allowEmptyValues) return fallbackValue;
+
+						throw new ArgumentNullException();
+					}
+
+					return (T)Enum.Parse(typeof(T), s, ignoreCase);
+				}, allowEmptyValues){ }
+			}
+		}
+
+		public class Format
+		{
+			private List<string> orderedKeys;
+			private IDictionary<string, Type> format;
+
+			public string Header
+			{
+				get
+				{
+					var sb = new StringBuilder();
+					Output(sb);
+					return sb.ToString();
+				}
+			}
+
+			public Format(IDictionary<string, Type> format)
+			{
+				this.format = format;
+			}
+
+			public string[] GetFieldNames()
+			{
+				var fieldCount = format.Keys.Count;
+				var fieldNames = new string[fieldCount];
+				for (var i = 0; i < fieldCount; i++)
+				{
+					fieldNames[i] = GetFieldName(i);
+				}
+				return fieldNames;
+			}
+
+			internal string GetFieldName(int index)
+			{
+				if (orderedKeys != null)
+				{
+					return (orderedKeys[index]);
+				}
+				else
+				{
+					return (format.Keys.ToArray())[index];
```

Comment:
`format.ElemantAt(index)`

Reply:
I'd recommend againt pulling things out of dictionaries by index. It's not guaranteed to be the same order you added them (although in my experience it always is, I think it's just an implementation detail). Also why is orderedKeys possible to be null ?

Reply:
Just for reference from: https://msdn.microsoft.com/en-us/library/xfhwa508.aspx 

> For purposes of enumeration, each item in the dictionary is treated as a KeyValuePair<TKey, TValue> structure representing a value and its key. The order in which the items are returned is undefined.